File: dynamic_programming/top_down_rod_cutting.py
import sys


def cut_rod(prices, length):
    results = [-sys.maxsize for _ in range(length + 1)]

    return cut_rod_internal(prices, length, results)


def cut_rod_internal(prices, length, results):
    if results[length] > -sys.maxsize:
        return results[length]

    if length == 0:
        result = 0
    else:
        result = -sys.maxsize
        for i in range(1, length + 1):
            result = max(result, prices[i] + cut_rod_internal(prices, length - i, results))

    results[length] = result

    return result


if __name__ == '__main__':
    prices = [0, 1, 5, 8, 9, 17, 17, 20, 24, 30]

    for i in range(len(prices)):
        result = cut_rod(prices, i)

        print(f"length = {i} / result = {result}")

File: dynamic_programming/top_down_rod_cutting.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func cutRod(n int, prices []int) int {
	results := make([]int, n+1)
	for i := 0; i < len(results); i++ {
		results[i] = -math.MaxInt
	}

	return cutRodInternal(n, prices, results)
}

func cutRodInternal(n int, prices, results []int) int {
	if results[n] > -math.MaxInt {
		return results[n]
	}

	result := 0
	if n > 0 {
		for i := 1; i <= n; i++ {
			result = max(results[i], prices[i]+cutRodInternal(n-i, prices, results))
		}
	}

	results[n] = result

	return result
}

func main() {
	prices := []int{0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30}

	for i := 0; i < len(prices); i++ {
		result := cutRod(i, prices)

		fmt.Printf("length = %d / result = %d\n", i, result)
	}
}
